src/nvim/decoration.c

```diff
@@ -924,7 +924,7 @@ int decor_virt_lines(win_T *wp, linenr_T lnum, VirtLines *lines, TriState has_fo
   while (true) {
     MTKey mark = marktree_itr_current(itr);
     DecorVirtText *vt = mt_decor_virt(mark);
-    if (ns_in_win(mark.ns, wp)) {
+    if (!mt_invalid(mark) && ns_in_win(mark.ns, wp)) {
```

Comment:
While at it, I wonder if `vt` should be checked here first before the `ns_in_win()` check, as `ns_in_win()` is more costly.
```suggestion
    if (vt && !mt_invalid(mark) && ns_in_win(mark.ns, wp)) {
```
The same seems to apply to `decor_redraw_col()` too, where `ns_in_win()` is currently the first check, but it will be better for it to be the last.

Reply:
Oh, it seems that `mt_end()` was actually removed from this function in #27011, but it wasn't removed elsewhere.

Reply:
Hmm yeah because we are iterating with the `kMTMetaLines` filter, I would assume `mark.flags & MT_FLAG_DECOR_VIRT_LINES` to always be true. Presumably `mt_end()` is still relevant and should be checked?

Reply:
Oh, I didn't notice the filter.